### PR TITLE
[ZEPPELIN-2245] separate precode into JDBCInterpreter

### PR DESCRIPTION
**Pull request by tinkoff-dwh**
### What is this PR for?
Separate precode by prefix. Added the ability to set different precode for different data sources

### What type of PR is it?
Improvement

### What is the Jira issue?
https://issues.apache.org/jira/browse/ZEPPELIN-2245

### How should this be tested?
1. Set properties
 ```
default.password 	1
default.precode 	set search_path='test_path'
default.url 	jdbc:postgresql://localhost:5432/
default.user 	postgres
mysql.driver 	com.mysql.jdbc.Driver
mysql.password 	1
mysql.precode 	set @v=12
mysql.url 	jdbc:mysql://localhost:3306/
mysql.user 	root 
```
2. Run 
`show search_path`
3. Run
```
%jdbc(mysql)
select @v
```

### Questions:
* Does the licenses files need update? no
* Is there breaking changes for older versions? no
* Does this needs documentation? no

### CODE REVIEW

**Comment by tinkoff-dwh** [comment]
@felixcheung 
done

**Comment by felixcheung** [comment]
link to #2078

File: docs/interpreter/jdbc.md [hunk 1]
```diff
@@ -118,6 +118,11 @@ The JDBC interpreter properties are defined by default like below.
     <td>gpadmin</td>
     <td>The JDBC user name</td>
   </tr>
+  <tr>
+    <td>default.precode</td>
+    <td></td>
+    <td>Some SQL which executes while opening connection</td>
+  </tr>
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: could you add an example of using this?
perhaps in ` ## Binding JDBC interpter to notebook`? Or another place you find

Comment (tinkoff-dwh) [inline docs/interpreter/jdbc.md]: added example

File: docs/interpreter/jdbc.md [hunk 2]
```diff
@@ -222,6 +222,66 @@ SELECT name, country, performer
...
+    <td>set @v=12</td>
+  </tr>
+</table>
+
+##### Usage
+```sql
+%jdbc
+show search_path
+```
+```sql
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: I think you need an extra empty line after ```

File: docs/interpreter/jdbc.md [hunk 3]
```diff
@@ -221,6 +222,66 @@ SELECT name, country, performer
...
+You can set `precode` for each data source. Code runs once while opening the connection.
+
+##### Properties
+<table class="table-configuration">
+  <tr>
+    <th>Property Name</th>
+    <th>Value</th>
+  </tr>
+  <tr>
+    <td>default.driver</td>
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: hmm, should it have/use `default.` here?

Comment (felixcheung) [inline docs/interpreter/jdbc.md]: maybe just omit them in this example?

File: docs/interpreter/jdbc.md [hunk 4]
```diff
@@ -221,6 +222,66 @@ SELECT name, country, performer
 FROM demo.performers
 WHERE name='{{"{{performer=Sheryl Crow|Doof|Fanfarlo|Los Paranoia"}}}}'
 ```
+### Usage `precode`
+You can set `precode` for each data source. Code runs once while opening the connection.
+
```
Comment (felixcheung) [inline docs/interpreter/jdbc.md]: you might want to add a line here to say `for example, first setup properties to mysql...`

File: docs/interpreter/jdbc.md (hunk 3 above)
Comment (tinkoff-dwh) [inline docs/interpreter/jdbc.md]: yes, here is set two data sources, each one with a precode

**Comment by felixcheung** [comment]
thanks, I see the changes, it would be good if you could add more descriptions on what they are doing

**Comment by tinkoff-dwh** [comment]
@felixcheung 
what kind of description, PR or changes from last commit?

**Comment by felixcheung** [comment]
To elaborate, this is what the documentation says
```
### Usage `precode`
<-- this is a new section
 +You can set `precode` for each data source. Code runs once while opening the connection.
<--- good, there's a description

 +##### Properties
<-- here we start a subsection

 +<table class="table-configuration">
 +  <tr>
 +    <th>Property Name</th>
 +    <th>Value</th>
 +  </tr>
 +  <tr>
 +    <td>default.driver</td>
 +    <td>org.postgresql.Driver</td>
<-- there we have a table immediately...
<-- what is default., mysql.? why do we need this properties/table?
 +  <tr>
 +    <td>mysql.url</td>
 +    <td>jdbc:mysql://localhost:3306/</td>
 +  <tr>
 +    <td>mysql.precode</td>
 +    <td>set @v=12</td>
 +  </tr>

 +</table>
 +
 +##### Usage
 +```sql
 +%jdbc(mysql)
 +select @v
<-- can we say here we get in return v, which is set in the precode
```

**Comment by tinkoff-dwh** [comment]
CI green https://travis-ci.org/tinkoff-dwh/zeppelin/builds/210855494

**Comment by felixcheung** [comment]
@Leemoonsoo it looks like Jenkins is timing out again while Travis is actually passing (though taking a long time)

**Comment by felixcheung** [comment]
merging if no more comment

**Pull request closed without merging**